ESet/ESet.h

```diff
@@ -0,0 +1,548 @@
+    }
+
+    iterator next = pos;
+    iterator result = end();
+    key_type next_key;
+
+    if (++next != result) {
+        next_key = *next;
+    }
+
+    remove_node(pos);
+
+    if (next != result) {
+        result = find(next_key);
+    }
+
+    return result;
+}
+
+template <typename K>
+typename ESet<K>::iterator
+ESet<K>::find(const key_type& key) const
+{
+    node_pointer_type node = root;
+
+    while (node != nullptr) {
+        if (key < node->data) {
+            node = node->get_left();
+        } else if (key > node->data) {
+            node = node->get_right();
+        } else {
+            return iterator(node, this);
+        }
+    }
+
+    return end();
+}
+
+template <typename K>
+bool ESet<K>::contains(const key_type& key) const
+{
+    return find(key) != end();
+}
+
+template <typename K>
+typename ESet<K>::node_pointer_type
+ESet<K>::min_node(node_pointer_type node) const
+{
+    while (node && node->has_left()) {
+        node = node->get_left();
+    }
+
+    return node;
+}
+
+template <typename K>
+void ESet<K>::remove_node(iterator pos)
+{
+    node_pointer_type removed_node = pos.current;
+
+    if (!removed_node->has_left() && !removed_node->has_right()) {
+        correct_pointers(removed_node, nullptr);
+    } else if (!removed_node->has_left() || !removed_node->has_right()) {
+        node_pointer_type child_node = removed_node->has_left() ? removed_node->get_left() : removed_node->get_right();
+
+        correct_pointers(removed_node, child_node);
+        child_node->set_parent(removed_node->get_parent());
+    } else {
+        node_pointer_type moved_node = min_node(removed_node->get_right());
+        removed_node->data = moved_node->data;
+
+        if (moved_node == removed_node->get_right()) {
+            removed_node->set_right(moved_node->get_right());
+        } else {
+            moved_node->get_parent()->set_left(moved_node->get_right());
+        }
+
+        if (moved_node->has_right()) {
+            moved_node->get_right()->set_parent(moved_node->get_parent());
+        }
+
+        removed_node = moved_node;
+    }
+
+    delete removed_node;
+    length--;
+}
+
+template <typename K>
+void ESet<K>::correct_pointers(node_pointer_type node, node_pointer_type new_node)
+{
+    if (!node->has_parent()) {
+        root = new_node;
+    } else if (node == node->get_parent()->get_left()) {
+        node->get_parent()->set_left(new_node);
+    } else {
+        node->get_parent()->set_right(new_node);
+    }
+}
+
+template <typename K>
+void ESet<K>::clean_node(node_pointer_type node)
+{
+    if (node != nullptr) {
+        clean_node(node->get_left());
+        clean_node(node->get_right());
+
+        delete node;
+    }
+}
+
+template <typename K>
+void ESet<K>::copy_tree(node_pointer_type node)
+{
+    if (node == nullptr) {
+        return;
+    }
+
+    insert(node->data);
+    copy_tree(node->get_left());
+    copy_tree(node->get_right());
```

Comment:
// Prerequisite: root should be allocated before calling this function
if (node == nullptr) {
   return;
}

if (!root)
{
root = new Node();
}

root->data = node->data;
copy_tree(root->left, node->left);
copy_tree(root->right, node->right);